tesla: workaround for north america single phase

thekev opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/7599

Ignore incorrect values in TWC3 vitals API when `singlephase: true` is set in configuration, use only the A phase and total vehicle current and return values as a single phase charger. Hopefully my translation to the German is close enough.

thekev committed:
tesla: workaround for north america single phase

andig commented:
Thank you @thekev. Are you writing from North America? Would be great if evcc did make it that far :)

Can we avoid the additional setting and the case switch between vehicle and "normal" current?

thekev commented:
> Thank you @thekev. Are you writing from North America? Would be great if evcc did make it that far :)
Yep. Utah, USA. Tesla EV, 2x Powerwalls, and 19.2kW Solar. A second EV is expected this year, so I’ll also be looking for EVCC to manage two vehicles and load points in the future. 

> Can we avoid the additional setting and the case switch between vehicle and "normal" current? Any way to auto-detect the situation, maybe using the 2-phase voltage supply?
I think I actually can. The anomalous reading on phase C (might be L1 to ground) is a likely indicator I expect you won’t see with true three phase service. I’ll give that a try. Thanks for the feedback.

andig commented:
Good idea- lets take the 120V +/- 10% or similar as an indicator?

thekev commented:
> Good idea- lets take the 120V +/- 10% or similar as an indicator?

Something like that. I think I’d compare it to the other phases and 45-55% as the condition to assume it’s the NA version. 

The amperage per phase is also not making sense. Where is that current on nonexistent phase C coming from? If vehicle reading is at 30A, and the sum of two wired phases is ~20A, but the sum of all three is 35A… I don’t know. Something in here has to make the math work. I’ll do some experiments. 

Do you want to avoid using the vehicle number because it may be representing the DC charge rate, and not the AC side (conversion losses)? I’m not clear why there’s a separate reading for that.

andig commented:
> Do you want to avoid using the vehicle number because it may be representing the DC charge rate, and not the AC side (conversion losses)? I’m not clear why there’s a separate reading for that.

I don't care really. But with https://github.com/evcc-io/evcc/pull/6897 we want to move apparent power calculation to the loadpoint instead of per-meter. To do that we need "real" currents from the chargers, not fake values. Would be great to also sanitise the currents for that purpose.

thekev commented:
@andig I refactored it to remove the configuration directive, instead using the single phase workaround based on the anomalous voltage seen on phase C. I think that should remain compatible with three phase units. I also adjusted the calculation for power to consider each phase's voltage in deriving apparent power.

andig commented on charger/twc3.go:
```diff
@@ -161,12 +163,32 @@ func (v *Twc3) ChargingTime() (time.Duration, error) {
+	//   "voltageB_v": 241.5,
+	//   "voltageC_v": 118.7,
+	avgAB := (res.VoltageAV + res.VoltageBV) / 2
+	if res.ContactorClosed && res.VoltageCV > avgAB / 2.1 && res.VoltageCV < avgAB / 1.9 {
```
Simplify as

    math.Abs(res.VoltageCV - avgAB / 2) < 25 // V

Why do we need the `ContactorClosed` here- voltages should always apply?

thekev replied:
When contactor is open, api reports voltage on all three phases as ~2V. These are measuring the output. I used contactor state to avoid considering the workaround when it can't yet be determined. However, by using your simplified math, and comparing phase C to the supply voltage field, the logic still works.

andig committed:
Simplify

andig commented:
Updated pr, let's see if it passes.

thekev commented:
> Updated pr, let's see if it passes.

Thanks for cleaning that up. My golang is not strong.

andig commented:
Are you happy with how it works? Then I could merge it!

thekev commented:
> Are you happy with how it works? Then I could merge it!

Looks good to me...other than the tesla api library is now broken, for unrelated reasons. Thanks for [working on that](https://github.com/bogosj/tesla/pull/98) too.
